### ISSUE TITLE
AmputatorBot doesn't work with thelocal.it pages

### ISSUE THREAD

**Issue by LeLobster** [KilledMufasa/AmputatorBot#7]
```
AmputatorBot couldn't reply to the comment or submission you summoned it for.

AmputatorBot ran into the following error: there were no canonical URLs found.

This error has been logged and is being investigated. Common causes for this error are: bot- and geoblocking websites and badly implemented AMP specs.

Feel free to leave feedback by contacting u/killed_mufasa, by posting on r/AmputatorBot or by opening an issue on GitHub.

You're a very good human for trying <3

NEW: With AmputatorBot.com you can remove AMP from your URLs in just one click! You could try it again there but it will probably raise an error again: https://AmputatorBot.com/?https://www.google.com/amp/s/www.thelocal.it/20191115/five-italian-police-officers-jailed-over-death-of-stefano-cucchi/amp

```

The url

https://www.google.com/amp/s/www.thelocal.it/20191115/five-italian-police-officers-jailed-over-death-of-stefano-cucchi/amp

**Comment by KilledMufasa** [comment]
Thanks for submitting this report. I have looked at the logs and AmputatorBot has failed 5 out 5 times in finding the canonical on thelocal.it websites. To replicate:

- www.amputatorbot.com/?https://www.thelocal.it/20200219/eating-well-driving-badly-and-daily-naps-the-habits-you-pick-up-in-italy/amp (non-cached)
- www.amputatorbot.com/?https://www.google.com/amp/s/www.thelocal.it/20180717/archaeological-enigma-rome-tiber-remains-church-villa/amp (cached)

It might have something to do with the fact that they've implemented Google's Shield for DDoS-protection. I'll let you know once I know more.

**Comment by KilledMufasa** [comment]
Hi again, bad news, the issue is indeed using Alphabet's Project Shield, they're even on their landing page: https://projectshield.withgoogle.com/landing?hl=en

Because of Project Shield, the scraper of AmputatorBot is not able to scrape the correct website, all it 'sees' is this: https://outline.com/4bKJLD - "You will be connected to www.thelocal.it in just a moment..." But since the bot can't pass the JavaScript challenge, let alone the other things we can't know about, AmputatorBot is blocked out.

Although there are some possible work-arounds to this issue, such as implementing a virtual browser, it's highly unlikely that these will work. Because well, it's meant to be hard to circumvent these kind of measures. Also, "Project Shield is a free service that defends news, human rights and election monitoring sites from DDoS attacks", trying to circumvent their DDoS-shield is, in my opinion, unethical, since it was probably put in place for good reason. Another way would be hardcoding a method, but that would be unmaintainable and since these are edge cases, I've decided not to go that route.

Instead, I've opted to keep better track of the websites where AmputatorBot doesn't work and give more detailed error-messages to reflect that we are aware of the issues with the website. For example, [this](https://www.amputatorbot.com/?https://www.google.com/amp/s/www.thelocal.it/20191115/five-italian-police-officers-jailed-over-death-of-stefano-cucchi/amp) will say something like this:

> Couldn't scrape this page and thus couldn't find the canonical link. It's probably blocking scraping services. This is a known issue specific to this domain.

This change is now in production: https://github.com/KilledMufasa/AmputatorBot/commit/3704e6405e3252866072a70c50e1d086447b0e6e

It's not the kind of solution either of us would like, in fact it isn't a solution at all, but I hope to reduce the amount of confusion with this change. Thx a lot for the bug-report!

**Comment by LeLobster** [comment]
Ah, fair enough. I personally don't see circumventing a DDOS shield for a one time scrape as that unethical but I understand your pov.  
is in those edge cases returning just the cleaned url without checking not an option?  
I understand that you want to check if the cleaned url is functional but isn't in most cases just removing the amp bits enough?
Thank you for your work!

**Comment by KilledMufasa** [comment]
Hi! Yeah circumventing is not _that_ unethical, by far the biggest issue is indeed that it is impossible to guess a canonical link and be sure that it's right. Your idea, to remove the amp bits, works sometimes. Other times, it becomes a 404. That's something I want to prevent of course. 

But this issue is something I've been annoyed by for quite some time. I've got a couple of other ideas:

- just guess, but find another way to check if it is the right link. For instance, I tried with outline, but often it just runs into the same issues as the bot: https://outline.com/4bKJLD
- just guess, don't check, but send a message when summoned that the results might not be entirely correct (this one probably makes the most sense)
- and by far the most ambitious one: use machine-learning :) - No wait hear me out, I've gathered about 10.000 link conversions now, and they tend to follow the same patterns we talked about earlier. Honestly I doubt this will work (because if one page doesn't work on a page, others probably won't either - so there are no correct examples for the problematic domains), but I could possibly find the correct links manually and feed it to the algorithm. Is this way too complex of a solution for such a minor issue? Absolutely, but I've been interested in machine learning for a while now, and this seems like a great experiment for the spring-vacation haha

I'm very busy at the moment, but in a couple of weeks I'll have plenty of time to see what the best way to deal with this problem is. Thanks for the ideas!

**Comment by LeLobster** [comment]
Do you have an example of an AMP link which errors when you remove the AMP? I don't think I've ever came across one myself.  

But yes, I think I agree that your second is probably the best solution, but the third is worth a shot if you're interested in ML  
I'd have no idea how to approach that so I'll wish you the best of luck if you do attempt it ha. Very ambitious

**Comment by KilledMufasa** [comment]
Shoot, totally forgot to respond, mb! I even mentioned this issue in the announcement post here: https://www.reddit.com/r/AmputatorBot/comments/i1wruh/amputatorbot_v30_better_bot/ 🤦 

I've been experimenting with your idea, and called it guess-and-check, here's what I wrote about it:

> Guess-and-check: Users occasionally suggest just trimming of the amp parts of URLs. With very variable results. Sometimes, it does the trick, but way too often it just breaks the page. To counter this, I've added some stuff to check the similarity between articles. If the articles are similar, we can say with some certainty that the guessed canonical is correct. If the canonical contains a rel=amphtml tag that points to the original URL, we know it for certain. That's the idea, but the results are simply so over the place that I'm not comfortable with enabling it elsewhere other than in mentions and online. Also, it's an extremely heavy CPU task because you can make a lot of guesses :p I'll continue experimenting, but I don't expect this method to be implemented fully any time soon.

As a test, I took the last 50 AMP URLs AmputatorBot handled and the guess-and-check results were as follows:
- Successful: 32
- Errors (wrong canonical): 13
- Failed: 5

At first glance, that might seem like an amazing result, and it is - but then consider the combined results of the other methods:
- Successful: 48
- Failed: 2

Of the 3 failed ones, guess-and-check is able to find 1 correct canonical and 1 wrong canoncial (passes criteria, but is e.g. not the same article as the AMP link). 

To summarize, the guess-and-check method can work, but doesn't really add much to the existing methods and it's an extremely CPU heavy task compared to the existing methods. It finds canonicals, but with an error rate of 26%. I would rather have the bot error out rather than spreading wrong links. With all that in mind, I'm gonna stick with the current solution, that is only use it if no other method gets results, and only online and when summoned. 

Thx again for the ideas!

**Comment by LeLobster** [comment]
Thanks for the nicely detailed write up, and v3.0 looks like a fantastic improvement to the bot  
Definitely agree that 'spreading wrong links' is unwanted, and i can fully get behind your reasoning for using the current methods    

Also something that Ive been wondering about; do you have an estimate of how many amp urls fail to be converted by normal methods?

**Comment by KilledMufasa** [comment]
Glad to hear you like it!

> Also something that Ive been wondering about; do you have an estimate of how many amp urls fail to be converted by normal methods? Im thinking it's not that big of a %

You're right, like I wrote:
> the combined results of the other methods:

    Successful: 48
    Failed: 2

I might have worded it badly, but those are the results of all methods except for guess-and-check. So we're looking at a success-rate of 95%-ish. Could be higher, but it's getting harder and harder to do so :/

**Comment by LeLobster** [comment]
No that was my bad. had a long day yesterday and I think I've just misinterpreted your comment, apologies    

It would of course be great to achieve a 100% success rate but if that means having to account for a handful of different edge cases through cpu heavy tasks then I think having >90% is already very nice  

AmputatorBot is doing a great job raising awareness, and Ive even seen some subreddits already "banning" amp urls altogether    
And one final question I have if that's ok, something else Ive been wondering  
Roughly estimated, how many requests does the bot handle per day?

**Comment by KilledMufasa** [comment]
No problem, I hope you're doing okay. 

You're right, I feel there has definitely been a change of narrative lately, it's great to see.
Here's the data of yesterday (07-08-2020):
- Online: 15
- Mention: 6
- Comment: 170
- Submission: 75
- TOTAL: 266

It was a bit of a slow day ¯\_(ツ)_/¯ normally it's closer to 300-350
That said, the above includes only AMP URLs. AmputatorBot needs to scan thousands of links (if not more, I don't have the data of this) for AMP. Perhaps a more accurate number, I use about 1100 CPU-seconds each day, that's up from about 700 with v2.2. 

Does that answer your question a bit?

**Comment by LeLobster** [comment]
I love numbers so thanks for sharing haha  
Considering its not allowed on some subs, still i'd actually expected the number to be way higher, but apparently it's not already as widespread as I thought  

Ive always been interested in how a reddit bot would work so I think I'll have a browse through your source code some time  
I use python myself too so maybe I'll pick up on some interesting tricks as well  

Thanks for all your work!

**Comment by KilledMufasa** [comment]
Np, I like numbers too. I gotta agree with you tho, it's not at much as I thought it would be on a per-day basis. 

Reddit bots are pretty cool, and Python is a great language to code in imo. V3.0 is object orientated, so if you find that complicated (I know I did) you might want to jump back a few commits, I think it's a bit easier to grasp. But you do you! You can check out this page for some inspiration on bots: b0trank.com

Nice talking to you, and good luck!
